File: trifecta-android/app/src/main/java/com/belweave/trifecta/designsystem/MarkdownText.kt
package com.belweave.trifecta.designsystem

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

/** Mobile-friendly markdown renderer mirroring iOS [MarkdownText]. */
@Composable
fun MarkdownText(
    source: String,
    baseStyle: TextStyle = T3Typography.body,
    secondaryColor: Color = T3Color.textSecondary,
    inlineCodeBackground: Color = T3Color.surfaceMuted,
    modifier: Modifier = Modifier
) {
    val blocks = remember(source) { MarkdownBlockParser.parse(source) }
    Column(modifier = modifier, verticalArrangement = Arrangement.spacedBy(T3Spacing.sm)) {
        blocks.forEach { block ->
            BlockView(
                block = block,
                baseStyle = baseStyle,
                secondaryColor = secondaryColor,
                inlineCodeBackground = inlineCodeBackground
            )
        }
    }
}

@Composable
private fun BlockView(
    block: MdBlock,
    baseStyle: TextStyle,
    secondaryColor: Color,
    inlineCodeBackground: Color
) {
    when (block) {
        is MdBlock.Paragraph -> Text(
            text = InlineMarkdown.parse(block.text, inlineCodeBackground),
            style = baseStyle,
            color = T3Color.textPrimary,
            modifier = Modifier.fillMaxWidth()
        )
        is MdBlock.Heading -> Text(
            text = InlineMarkdown.parse(block.text, inlineCodeBackground, boldAll = true),
            style = headingStyle(block.level),
            color = T3Color.textPrimary,
            modifier = Modifier.fillMaxWidth()
        )
        is MdBlock.Code -> CodeBlock(code = block.code, language = block.language)
        is MdBlock.Bullet -> ListBlock(
            items = block.items,
            ordered = false,
            baseStyle = baseStyle,
            inlineBackground = inlineCodeBackground
        )
        is MdBlock.Numbered -> ListBlock(
            items = block.items,
            ordered = true,
            baseStyle = baseStyle,
            inlineBackground = inlineCodeBackground
        )
        is MdBlock.Quote -> QuoteBlock(
            lines = block.lines,
            baseStyle = baseStyle,
            inlineBackground = inlineCodeBackground,
            textColor = secondaryColor
        )
        MdBlock.Divider -> Box(
            modifier = Modifier
                .fillMaxWidth()
                .padding(vertical = T3Spacing.xs)
                .background(T3Color.separator)
        ) { Box(modifier = Modifier.size(0.dp, 0.5.dp)) }
    }
}

@Composable
private fun headingStyle(level: Int): TextStyle = when (level) {
    1 -> T3Typography.title
    2 -> T3Typography.headline
    3 -> T3Typography.bodyEmphasis
    else -> T3Typography.callout.copy(fontWeight = FontWeight.SemiBold)
}

@Composable
private fun ListBlock(
    items: List<MdListItem>,
    ordered: Boolean,
    baseStyle: TextStyle,
    inlineBackground: Color
) {
    Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
        items.forEachIndexed { index, item ->
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(start = (item.depth * 16).dp),
                verticalAlignment = Alignment.Top,
                horizontalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                Text(
                    text = if (ordered) "${index + 1}." else "•",
                    style = baseStyle.copy(fontWeight = if (ordered) FontWeight.SemiBold else FontWeight.Normal),
                    color = T3Color.textSecondary,
                    modifier = Modifier.width(if (ordered) 22.dp else 14.dp)
                )
                Text(
                    text = InlineMarkdown.parse(item.text, inlineBackground),
                    style = baseStyle,
                    color = T3Color.textPrimary,
                    modifier = Modifier.fillMaxWidth()
                )
            }
        }
    }
}

@Composable
private fun QuoteBlock(
    lines: List<String>,
    baseStyle: TextStyle,
    inlineBackground: Color,
    textColor: Color
) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(vertical = T3Spacing.xs),
        horizontalArrangement = Arrangement.spacedBy(T3Spacing.sm)
    ) {
        Box(modifier = Modifier.width(3.dp).background(T3Color.separator)) {
            Box(modifier = Modifier.size(3.dp, 1.dp))
        }
        Text(
            text = InlineMarkdown.parse(lines.joinToString("\n"), inlineBackground),
            style = baseStyle,
            color = textColor,
            modifier = Modifier.fillMaxWidth()
        )
    }
}

@Composable
fun CodeBlock(code: String, language: String?) {
    val isDiff = language?.lowercase() in setOf("diff", "patch", "udiff", "git")
    Box(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(T3Radius.md))
            .background(T3Color.surfaceMuted)
            .border(0.5.dp, T3Color.separator, RoundedCornerShape(T3Radius.md))
            .padding(horizontal = T3Spacing.md, vertical = T3Spacing.sm)
    ) {
        if (isDiff) {
            Column {
                code.split("\n").forEach { line ->
                    Text(
                        text = line,
                        style = T3Typography.code,
                        color = diffLineColor(line),
                        modifier = Modifier.fillMaxWidth()
                    )
                }
            }
        } else {
            Text(
                text = code,
                style = T3Typography.code,
                color = T3Color.textPrimary,
                modifier = Modifier.fillMaxWidth()
            )
        }
    }
}

@Composable
private fun diffLineColor(line: String): Color {
    val t = line.trim()
    if (t.startsWith("+++ ") || t.startsWith("--- ")) return T3Color.textSecondary
    if (t.startsWith("diff --git") || t.startsWith("Index: ")) return T3Color.textSecondary
    if (t.startsWith("@@")) return T3Color.warning
    return when (line.firstOrNull()) {
        '+' -> T3Color.success
        '-' -> T3Color.danger
        else -> T3Color.textPrimary
    }
}

// region Block model

internal sealed class MdBlock {
    data class Paragraph(val text: String) : MdBlock()
    data class Heading(val level: Int, val text: String) : MdBlock()
    data class Code(val code: String, val language: String?) : MdBlock()
    data class Bullet(val items: List<MdListItem>) : MdBlock()
    data class Numbered(val items: List<MdListItem>) : MdBlock()
    data class Quote(val lines: List<String>) : MdBlock()
    object Divider : MdBlock()
}

internal data class MdListItem(val depth: Int, val text: String)

// endregion

// region Block parser

internal object MarkdownBlockParser {
    fun parse(source: String): List<MdBlock> {
        val blocks = mutableListOf<MdBlock>()
        val lines = source.split("\n")
        var i = 0
        while (i < lines.size) {
            val line = lines[i]
            val trimmed = line.trim()

            if (trimmed.startsWith("```")) {
                val langCandidate = trimmed.removePrefix("```").trim()
                val language = langCandidate.takeIf { it.isNotEmpty() }
                i += 1
                val code = StringBuilder()
                var first = true
                while (i < lines.size) {
                    if (lines[i].trim().startsWith("```")) {
                        i += 1
                        break
                    }
                    if (!first) code.append('\n')
                    code.append(lines[i])
                    first = false
                    i += 1
                }
                blocks.add(MdBlock.Code(code.toString(), language))
                continue
            }

            val heading = parseHeading(trimmed)
            if (heading != null) {
                blocks.add(MdBlock.Heading(heading.first, heading.second))
                i += 1
                continue
            }

            if (isHorizontalRule(trimmed)) {
                blocks.add(MdBlock.Divider)
                i += 1
                continue
            }

            if (trimmed.startsWith(">")) {
                val q = mutableListOf<String>()
                while (i < lines.size) {
                    val lt = lines[i].trim()
                    if (!lt.startsWith(">")) break
                    var content = lt.removePrefix(">")
                    if (content.startsWith(" ")) content = content.substring(1)
                    q.add(content)
                    i += 1
                }
                blocks.add(MdBlock.Quote(q))
                continue
            }

            val kind = listKind(line)
            if (kind != null) {
                val items = mutableListOf<MdListItem>()
                while (i < lines.size && listKind(lines[i]) == kind) {
                    items.add(parseListItem(lines[i]))
                    i += 1
                }
                blocks.add(if (kind == ListKind.BULLET) MdBlock.Bullet(items) else MdBlock.Numbered(items))
                continue
            }

            if (trimmed.isEmpty()) {
                i += 1
                continue
            }

            val paraLines = mutableListOf<String>()
            while (i < lines.size) {
                val raw = lines[i]
                val rt = raw.trim()
                if (rt.isEmpty()) break
                if (rt.startsWith("```")) break
                if (isHorizontalRule(rt)) break
                if (parseHeading(rt) != null) break
                if (rt.startsWith(">")) break
                if (listKind(raw) != null) break
                paraLines.add(raw)
                i += 1
            }
            if (paraLines.isNotEmpty()) {
                blocks.add(MdBlock.Paragraph(paraLines.joinToString("\n")))
            }
        }
        return blocks
    }

    private fun parseHeading(trimmed: String): Pair<Int, String>? {
        if (!trimmed.startsWith("#")) return null
        var level = 0
        var rest = trimmed
        while (rest.startsWith("#") && level < 6) {
            rest = rest.substring(1)
            level += 1
        }
        if (level == 0 || !rest.startsWith(" ")) return null
        val text = rest.trim()
        if (text.isEmpty()) return null
        return level to text
    }

    private fun isHorizontalRule(trimmed: String): Boolean {
        if (trimmed.length < 3) return false
        val chars = trimmed.toSet()
        return chars == setOf('-') || chars == setOf('*') || chars == setOf('_')
    }

    private enum class ListKind { BULLET, NUMBERED }

    private fun listKind(raw: String): ListKind? {
        val t = raw.trim()
        if (t.startsWith("- ") || t.startsWith("* ")) return ListKind.BULLET
        val dot = t.indexOf('.')
        if (dot > 0 && t.substring(0, dot).all { it.isDigit() } && dot + 1 < t.length && t[dot + 1] == ' ') {
            return ListKind.NUMBERED
        }
        return null
    }

    private fun parseListItem(line: String): MdListItem {
        var leading = 0
        for (c in line) {
            when (c) {
                ' ' -> leading += 1
                '\t' -> leading += 4
                else -> break
            }
        }
        val depth = leading / 2
        var text = line.trim()
        if (text.isNotEmpty() && (text[0] == '-' || text[0] == '*') && text.length >= 2 && text[1] == ' ') {
            text = text.substring(2)
        } else {
            val dot = text.indexOf('.')
            if (dot > 0 && text.substring(0, dot).all { it.isDigit() } && dot + 1 < text.length && text[dot + 1] == ' ') {
                text = text.substring(dot + 2)
            }
        }
        return MdListItem(depth, text)
    }
}

// endregion

// region Inline parser

internal object InlineMarkdown {
    fun parse(source: String, codeBackground: Color, boldAll: Boolean = false): AnnotatedString =
        buildAnnotatedString {
            val pushBase = if (boldAll) {
                pushStyle(SpanStyle(fontWeight = FontWeight.Bold))
                true
            } else false

            var i = 0
            while (i < source.length) {
                val c = source[i]
                when {
                    c == '\\' && i + 1 < source.length -> {
                        append(source[i + 1])
                        i += 2
                    }
                    c == '`' -> {
                        val end = source.indexOf('`', i + 1)
                        if (end > 0) {
                            val code = source.substring(i + 1, end)
                            pushStyle(SpanStyle(background = codeBackground, fontFamily = FontFamily.Monospace, fontSize = 13.sp))
                            append(code)
                            pop()
                            i = end + 1
                        } else {
                            append(c); i += 1
                        }
                    }
                    c == '*' || c == '_' -> {
                        val isBold = i + 1 < source.length && source[i + 1] == c
                        val marker = if (isBold) source.substring(i, i + 2) else c.toString()
                        val searchFrom = i + marker.length
                        val end = source.indexOf(marker, searchFrom)
                        if (end > 0) {
                            val text = source.substring(searchFrom, end)
                            val style = if (isBold) SpanStyle(fontWeight = FontWeight.Bold)
                                        else SpanStyle(fontStyle = FontStyle.Italic)
                            pushStyle(style)
                            append(text)
                            pop()
                            i = end + marker.length
                        } else {
                            append(c); i += 1
                        }
                    }
                    c == '[' -> {
                        val close = source.indexOf(']', i + 1)
                        if (close > 0 && close + 1 < source.length && source[close + 1] == '(') {
                            val parenClose = source.indexOf(')', close + 2)
                            if (parenClose > 0) {
                                val linkText = source.substring(i + 1, close)
                                pushStyle(SpanStyle(color = Color(0xFF4F6BED)))
                                append(linkText)
                                pop()
                                i = parenClose + 1
                                continue
                            }
                        }
                        append(c); i += 1
                    }
                    else -> {
                        append(c); i += 1
                    }
                }
            }

            if (pushBase) pop()
        }
}
// endregion

File: trifecta-ios/Trifecta/DesignSystem/Components/MarkdownText.swift
import SwiftUI

/// Mobile-friendly block-level markdown renderer.
///
/// Supports the markdown subset assistants typically emit:
/// - Fenced code blocks (```lang … ```)
/// - ATX headings (`#` through `######`)
/// - Bullet lists (`-`, `*`, `+`)
/// - Numbered lists (`1.`, `2.`, …)
/// - Block quotes (`> …`)
/// - Horizontal rules (`---`, `***`, `___`)
/// - Paragraphs separated by blank lines
///
/// Inline markdown (bold, italic, inline code, links) is rendered via Apple's
/// `AttributedString(markdown:)` so SwiftUI handles wrapping/selection natively.
struct MarkdownText: View {
    let source: String
    var baseFont: Font = T3Typography.body
    var secondaryColor: Color = T3Color.textSecondary
    var inlineCodeBackground: Color = T3Color.surfaceMuted

    var body: some View {
        VStack(alignment: .leading, spacing: T3Spacing.sm) {
            ForEach(Array(blocks.enumerated()), id: \.offset) { _, block in
                blockView(block)
            }
        }
    }

    private var blocks: [Block] {
        MarkdownBlockParser.parse(source)
    }

    @ViewBuilder
    private func blockView(_ block: Block) -> some View {
        switch block {
        case .paragraph(let text):
            paragraphText(text)
        case .heading(let level, let text):
            paragraphText(text, font: headingFont(for: level), bold: true)
                .padding(.top, level <= 2 ? T3Spacing.xs : 0)
        case .code(let code, let language):
            CodeBlockView(code: code, language: language)
        case .bullet(let items):
            ListBlockView(items: items,
                          ordered: false,
                          baseFont: baseFont,
                          inlineBackground: inlineCodeBackground)
        case .numbered(let items):
            ListBlockView(items: items,
                          ordered: true,
                          baseFont: baseFont,
                          inlineBackground: inlineCodeBackground)
        case .quote(let lines):
            QuoteBlockView(lines: lines,
                           baseFont: baseFont,
                           inlineBackground: inlineCodeBackground,
                           textColor: secondaryColor)
        case .divider:
            Rectangle()
                .fill(T3Color.separator)
                .frame(height: 0.5)
                .padding(.vertical, T3Spacing.xs)
        }
    }

    private func paragraphText(_ source: String,
                               font: Font? = nil,
                               bold: Bool = false) -> some View {
        Text(InlineMarkdown.attributed(source,
                                       inlineCodeBackground: inlineCodeBackground,
                                       boldAll: bold))
            .font(font ?? baseFont)
            .foregroundStyle(T3Color.textPrimary)
            .lineSpacing(3)
            .textSelection(.enabled)
            .frame(maxWidth: .infinity, alignment: .leading)
    }

    private func headingFont(for level: Int) -> Font {
        switch level {
        case 1: return T3Typography.title
        case 2: return T3Typography.headline
        case 3: return T3Typography.bodyEmphasis
        default: return T3Typography.callout
        }
    }

    // MARK: - Block model

    enum Block {
        case paragraph(String)
        case heading(Int, String)
        case code(String, String?)
        case bullet([ListItem])
        case numbered([ListItem])
        case quote([String])
        case divider
    }

    struct ListItem {
        let depth: Int
        let text: String
    }
}

// MARK: - Inline rendering

private enum InlineMarkdown {
    static func attributed(_ source: String,
                           inlineCodeBackground: Color,
                           boldAll: Bool = false) -> AttributedString {
        let mdOptions = AttributedString.MarkdownParsingOptions(
            interpretedSyntax: .inlineOnlyPreservingWhitespace,
            failurePolicy: .returnPartiallyParsedIfPossible
        )
        guard var attributed = try? AttributedString(markdown: source, options: mdOptions) else {
            var fallback = AttributedString(source)
            if boldAll { fallback.font = .body.bold() }
            return fallback
        }
        for run in attributed.runs {
            guard let intent = run.inlinePresentationIntent,
                  intent.contains(.code) else { continue }
            attributed[run.range].backgroundColor = inlineCodeBackground
            attributed[run.range].foregroundColor = T3Color.textPrimary
            attributed[run.range].font = .system(.footnote, design: .monospaced).weight(.medium)
        }
        return attributed
    }
}

// MARK: - List & quote helpers

private struct ListBlockView: View {
    let items: [MarkdownText.ListItem]
    let ordered: Bool
    let baseFont: Font
    let inlineBackground: Color

    var body: some View {
        VStack(alignment: .leading, spacing: 4) {
            ForEach(Array(items.enumerated()), id: \.offset) { index, item in
                HStack(alignment: .firstTextBaseline, spacing: 8) {
                    marker(index: index)
                        .font(baseFont)
                        .foregroundStyle(T3Color.textSecondary)
                        .frame(minWidth: ordered ? 22 : 14, alignment: .trailing)
                    Text(InlineMarkdown.attributed(item.text,
                                                   inlineCodeBackground: inlineBackground))
                        .font(baseFont)
                        .foregroundStyle(T3Color.textPrimary)
                        .lineSpacing(3)
                        .textSelection(.enabled)
                        .frame(maxWidth: .infinity, alignment: .leading)
                }
                .padding(.leading, CGFloat(item.depth) * 16)
            }
        }
    }

    @ViewBuilder
    private func marker(index: Int) -> some View {
        if ordered {
            Text("\(index + 1).").fontWeight(.semibold)
        } else {
            Text("•")
        }
    }
}

private struct QuoteBlockView: View {
    let lines: [String]
    let baseFont: Font
    let inlineBackground: Color
    let textColor: Color

    var body: some View {
        HStack(alignment: .top, spacing: T3Spacing.sm) {
            Rectangle()
                .fill(T3Color.separator)
                .frame(width: 3)
            Text(InlineMarkdown.attributed(lines.joined(separator: "\n"),
                                           inlineCodeBackground: inlineBackground))
                .font(baseFont)
                .foregroundStyle(textColor)
                .lineSpacing(3)
                .textSelection(.enabled)
                .frame(maxWidth: .infinity, alignment: .leading)
        }
        .padding(.vertical, T3Spacing.xs)
    }
}

// MARK: - Block parser

private enum MarkdownBlockParser {
    static func parse(_ source: String) -> [MarkdownText.Block] {
        var blocks: [MarkdownText.Block] = []
        let lines = source.components(separatedBy: "\n")
        var i = 0
        while i < lines.count {
            let line = lines[i]
            let trimmed = line.trimmingCharacters(in: .whitespaces)

            // Fenced code block
            if trimmed.hasPrefix("```") {
                let langCandidate = trimmed.dropFirst(3).trimmingCharacters(in: .whitespaces)
                let language = langCandidate.isEmpty ? nil : String(langCandidate)
                i += 1
                var codeLines: [String] = []
                while i < lines.count {
                    let l = lines[i]
                    if l.trimmingCharacters(in: .whitespaces).hasPrefix("```") {
                        i += 1
                        break
                    }
                    codeLines.append(l)
                    i += 1
                }
                blocks.append(.code(codeLines.joined(separator: "\n"), language))
                continue
            }

            // Heading
            if let heading = parseHeading(trimmed) {
                blocks.append(.heading(heading.level, heading.text))
                i += 1
                continue
            }

            // Horizontal rule
            if isHorizontalRule(trimmed) {
                blocks.append(.divider)
                i += 1
                continue
            }

            // Block quote (consume contiguous quote lines)
            if trimmed.hasPrefix(">") {
                var quoteLines: [String] = []
                while i < lines.count {
                    let lineTrimmed = lines[i].trimmingCharacters(in: .whitespaces)
                    if !lineTrimmed.hasPrefix(">") { break }
                    var content = lineTrimmed
                    content.removeFirst()
                    if content.hasPrefix(" ") { content.removeFirst() }
                    quoteLines.append(content)
                    i += 1
                }
                blocks.append(.quote(quoteLines))
                continue
            }

            // Unfenced unified diff / patch (common in assistant output without ```)
            if let diff = consumeUnfencedDiff(lines: lines, start: i) {
                blocks.append(.code(diff.text, "diff"))
                i = diff.end
                continue
            }

            // List
            if let kind = listKind(for: line) {
                var items: [MarkdownText.ListItem] = []
                while i < lines.count, let next = listKind(for: lines[i]), next == kind {
                    items.append(parseListItem(line: lines[i]))
                    i += 1
                }
                if kind == .bullet {
                    blocks.append(.bullet(items))
                } else {
                    blocks.append(.numbered(items))
                }
                continue
            }

            // Blank line
            if trimmed.isEmpty {
                i += 1
                continue
            }

            // Paragraph: collect contiguous non-empty, non-special lines
            var paraLines: [String] = []
            while i < lines.count {
                let raw = lines[i]
                let lineTrimmed = raw.trimmingCharacters(in: .whitespaces)
                if lineTrimmed.isEmpty { break }
                if lineTrimmed.hasPrefix("```") { break }
                if isHorizontalRule(lineTrimmed) { break }
                if parseHeading(lineTrimmed) != nil { break }
                if lineTrimmed.hasPrefix(">") { break }
                if listKind(for: raw) != nil { break }
                if isDiffStrongHeader(lineTrimmed) { break }
                if looksLikeWeakDiffRun(lines: lines, start: i) { break }
                paraLines.append(raw)
                i += 1
            }
            if !paraLines.isEmpty {
                blocks.append(.paragraph(paraLines.joined(separator: "\n")))
            }
        }
        return blocks
    }

    private static func parseHeading(_ trimmed: String) -> (level: Int, text: String)? {
        guard trimmed.hasPrefix("#") else { return nil }
        var level = 0
        var rest = Substring(trimmed)
        while rest.first == "#", level < 6 {
            rest = rest.dropFirst()
            level += 1
        }
        guard level > 0, rest.first == " " else { return nil }
        let text = rest.dropFirst().trimmingCharacters(in: .whitespaces)
        guard !text.isEmpty else { return nil }
        return (level, text)
    }

    private static func isHorizontalRule(_ trimmed: String) -> Bool {
        guard trimmed.count >= 3 else { return false }
        let chars = Set(trimmed)
        return chars == ["-"] || chars == ["*"] || chars == ["_"]
    }

    enum ListKind { case bullet, numbered }

    private static func listKind(for raw: String) -> ListKind? {
        let trimmed = raw.trimmingCharacters(in: .whitespaces)
        // Omit `+ ` — it collides with unified-diff added lines in streamed output.
        if trimmed.hasPrefix("- ") || trimmed.hasPrefix("* ") {
            return .bullet
        }
        if let dot = trimmed.firstIndex(of: ".") {
            let prefix = trimmed[..<dot]
            if !prefix.isEmpty,
               prefix.allSatisfy({ $0.isASCII && $0.isNumber }) {
                let after = trimmed.index(after: dot)
                if after < trimmed.endIndex, trimmed[after] == " " {
                    return .numbered
                }
            }
        }
        return nil
    }

    private static func parseListItem(line: String) -> MarkdownText.ListItem {
        var leadingSpaces = 0
        for c in line {
            if c == " " { leadingSpaces += 1 }
            else if c == "\t" { leadingSpaces += 4 }
            else { break }
        }
        let depth = max(0, leadingSpaces / 2)
        var text = line.trimmingCharacters(in: .whitespaces)
        if let first = text.first, "-*".contains(first), text.count >= 2 {
            let afterMarker = text.index(after: text.startIndex)
            if text[afterMarker] == " " {
                text = String(text[text.index(after: afterMarker)...])
            }
        } else if let dot = text.firstIndex(of: ".") {
            let prefix = text[..<dot]
            if prefix.allSatisfy({ $0.isASCII && $0.isNumber }) {
                let after = text.index(after: dot)
                if after < text.endIndex, text[after] == " " {
                    text = String(text[text.index(after: after)...])
                }
            }
        }
        return MarkdownText.ListItem(depth: depth, text: text)
    }

    // MARK: - Unfenced diff

    /// Captures git-style patches when the model omits fenced code blocks so `-` / `+`
    /// lines are not mis-parsed as bullet lists and hunks stay monospace.
    private static func consumeUnfencedDiff(lines: [String], start: Int) -> (text: String, end: Int)? {
        guard start < lines.count else { return nil }
        let line = lines[start]
        let trimmed = line.trimmingCharacters(in: .whitespaces)

        if isDiffStrongHeader(trimmed) {
            return consumeDiffBody(lines: lines, start: start)
        }
        guard looksLikeWeakDiffRun(lines: lines, start: start) else { return nil }
        return consumeDiffBody(lines: lines, start: start)
    }

    private static func consumeDiffBody(lines: [String], start: Int) -> (text: String, end: Int)? {
        var j = start
        var collected: [String] = []
        while j < lines.count {
            let raw = lines[j]
            let t = raw.trimmingCharacters(in: .whitespaces)
            if t.isEmpty {
                if !collected.isEmpty { break }
                j += 1
                continue
            }
            if !isDiffBodyLine(raw) { break }
            collected.append(raw)
            j += 1
        }
        guard !collected.isEmpty else { return nil }
        return (collected.joined(separator: "\n"), j)
    }

    private static func isDiffStrongHeader(_ trimmed: String) -> Bool {
        if trimmed.hasPrefix("@@") { return true }
        if trimmed.hasPrefix("diff --git") { return true }
        if trimmed.hasPrefix("Index: ") { return true }
        if trimmed.hasPrefix("--- a/") || trimmed.hasPrefix("--- b/")
            || trimmed.hasPrefix("+++ a/") || trimmed.hasPrefix("+++ b/") {
            return true
        }
        return false
    }

    /// Weak detection: ≥2 lines that look like a unified hunk without file headers.
    private static func looksLikeWeakDiffRun(lines: [String], start: Int) -> Bool {
        guard start + 1 < lines.count else { return false }
        var i = start
        var sawPlus = false
        var sawMinus = false
        var allMinusSpaceBullets = true

        while i < lines.count {
            let raw = lines[i]
            if raw.trimmingCharacters(in: .whitespaces).isEmpty { break }
            guard let first = raw.first else { break }

            if first == "+" {
                sawPlus = true
                allMinusSpaceBullets = false
            } else if first == "-" {
                sawMinus = true
                let rest = raw.dropFirst()
                if rest.first != " " {
                    allMinusSpaceBullets = false
                }
            } else if first == " " || first == "\t" {
                allMinusSpaceBullets = false
            } else {
                break
            }
            i += 1
        }

        let count = i - start
        guard count >= 2 else { return false }
        if sawPlus { return true }
        if allMinusSpaceBullets { return false }
        return sawMinus
    }

    private static func isDiffBodyLine(_ raw: String) -> Bool {
        let t = raw.trimmingCharacters(in: .whitespaces)
        if t.hasPrefix("@@") || t.hasPrefix("diff --git") || t.hasPrefix("Index: ") {
            return true
        }
        if t.hasPrefix("+++ ") || t.hasPrefix("--- ") {
            return true
        }
        if t.hasPrefix("index ") && t.contains("..") {
            return true
        }
        if t == "\\ No newline at end of file" {
            return true
        }
        guard let first = raw.first else { return false }
        if first == "+" || first == "-" {
            return true
        }
        if first == " " || first == "\t" {
            return raw.count > 1
        }
        if first == "\\" {
            return true
        }
        return false
    }
}
